app/lib/service/download_counts/download_counts.dart

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:basics/basics.dart';
+import 'package:json_annotation/json_annotation.dart';
+import 'package:pub_semver/pub_semver.dart';
+part 'download_counts.g.dart';
+
+@JsonSerializable(includeIfNull: false)
+class CountData {
+  // We store at most two years of data, one data point per day.
+  static const countsCapacity = 731;
+  final versionCounts = <({String versionRange, List<int> counts})>[];
+  final patchCounts = <({String versionRange, List<int> counts})>[];
+  final minorCounts = <({String versionRange, List<int> counts})>[];
+  final majorCounts = <({String versionRange, List<int> counts})>[];
+  final totalCounts = <int>[];
+
+  // Last date with processed data.
+  // The date only contains year, month and date.
+  // Hours, minutes and seconds are disregarded.
+  DateTime? lastDate;
+
+  CountData();
+
+  void addDownloadCounts(Map<String, int> dayCounts, DateTime dateTime) {
+    final date = DateTime(dateTime.year, dateTime.month, dateTime.day);
+    lastDate ??= date.addCalendarDays(-1);
+
+    final nextLastDate = date.isAfter(lastDate!) ? date : lastDate!;
+    final countsIndex = nextLastDate.difference(date).inDays;
+
+    if (countsIndex >= countsCapacity) {
+      // We don't store counts that are more than two years old.
+      return;
+    }
+
+    _prepareDates(date, lastDate!, dayCounts, countsIndex, majorCounts);
+
+    _processCounts(
+        dayCounts, date, countsIndex, majorCounts, _createNewMajorVersionRange);
+
+    lastDate = nextLastDate;
+  }
+
+  void _prepareDates(
+    DateTime date,
+    DateTime workingLastDate,
+    Map<String, int> dayCounts,
+    int countsIndex,
+    List<({List<int> counts, String versionRange})> versionCounts,
+  ) {
+    if (date.isAfter(DateTime.now())) {
```

Comment:
```suggestion
    if (date.isAfter(clock.now())) {
```

for better testing

Reply:
also I'd suggest throwing an `ArgumentError`. Isn't it fair to call this a serious violation of our invariants. We probably want things to make noise.

Reply:
I removed this check as well. I can add it again later if we think it should be there.